### DISCUSSION TITLE
Tilemap : Add ability to separately scale tiles images or rotate them by specific degrees

### DISCUSSION THREAD

**Discussion by Xinart** [godotengine/godot-proposals#10290] [2D]
## Describe the project you are working on
2D Metroidvania game porototypes on Godot 4.3

### 1. Sprites/textures drawn from individual tiles of tilemaps cannot be rotated in specific degrees. Currently limited to the 90° ,180° ,270° ,360°

To create organic and interesting looking levels i need to rotate image of tile by precise degrees. I'm doing so by creating separate tilemap per degree but it is really not optimized, relative to center point of tilemap, and is hard to manage more i want diversity

I know this can be done separatly with sprite but i wanna use tilemaps beceause they are powerfull, allow me to work easily with spritesheets and are really speeding up my working process beceaus it is centralized around it
![S2K4DFkCoL](https://github.com/user-attachments/assets/d887d512-c14f-402c-9eb7-1668de58550c)

### 2. Sprites/textures drawn from individual tiles of tilemaps cannot be scaled.

To create organic and interesting looking levels i need to scale differently image of tile. I'm doing so by creating separate tilemap per scale but it is really not optimized and is hard to manage more i want diversity

For same reason in point 1 i wanna use tilemaps to do so

![HI8AmWzWbD](https://github.com/user-attachments/assets/4ac0057c-93de-4a4d-bf79-eb697b1dc9d2)

### Feature 1 : Tile image rotation by degrees
Adding an option to rotate the sprite/tecture of the selected tile in the tilemap editor
![image](https://github.com/user-attachments/assets/73d1e36c-13fb-4f41-abd0-ff9692ae0a47)

### Feature 2 : Tile image scaling
Adding an option to scale the sprite/tecture of the selected tile in the tilemap editor
![8HfNDpXZf5](https://github.com/user-attachments/assets/e581cb4a-fe9c-402c-8793-6cb7f4162e42)

### Feature 1 : Tile image rotation by degrees
Tweaking the _**Tile image rotation**_ property will rotate the related image of the selected tile and for the next one to draw
![image](https://github.com/user-attachments/assets/70d10b40-9fa7-4c05-b34d-162b778882f9)

### Feature 2 : Tile image scaling
Adding an option to scale the sprite/tecture of the selected tile in the tilemap editor
Tweaking the **_Tile Scale_** property will scale the related image of the selected tile and for the next one to draw
![image](https://github.com/user-attachments/assets/a7cc4ac9-c4f3-44b5-bcff-b84fb9736ccf)

## If this enhancement will not be used often, can it be worked around with a few lines of script?
I didn't found anything that can make it work in scripts

## Is there a reason why this should be core and not an add-on in the asset library?
This proposal is a direct extension of the tilemap features an may be usefull to a lot of other users, it make no sense  to create a separate add-on for it

**Comment by AThousandShips** [comment]
This should be a proper proposal rather than a discussion as it is sufficiently detailed, discussions are for discussing ideas without concrete ideas for how to accomplish it

**Reply by Xinart** [reply]
I saw in the [Godot Improvment Reference ](https://github.com/godotengine/godot-proposals#:~:text=If%20you%20have%20a%20more%20general%20idea%20for%20a%20feature%20but%20are%20not%20well%20versed%20in%20Godot%27s%20architecture%20or%20do%20not%20possess%20the%20necessary%20knowledge%20to%20implement%20it%20in%20the%20engine%2C%20feel%20free%20to%20open%20a%20discussion) That there is two way to make a suggestion and I ended with one of it with this discussion beceaus i'm not confident enough to implement this myself in the engine. Was it inapropriate ? And if yes how should i proceed to make it a proper proposal as you sayed ?

**Reply by AThousandShips** [reply]
You don't have to be able to implement it yourself to make a proposal, it's more about how detailed your description is, the detail in the description you linked is "more general idea for a feature", good to note that this part could be more clear

A proposal (as an issue, not a discussion) is easier to track and handles details better than discussion threads, I'd suggest just creating an issue with the full descriptions you have here, as you copied the way it is filled in

**Reply by Xinart** [reply]
Ok many thanls for the clarifications, i copied the content of this discussion and transferred to this issue : https://github.com/godotengine/godot-proposals/issues/10308

**Comment by KoBeWi** [comment]
You can use scene tiles for sprites and custom data on another layer to store rotation.
But you should be using Sprite2D for that. Due to how tile data is stored, there is not much space to add rotation. The current "tile rotation" reserves 3 bits from alternate tile ID.

**Reply by Xinart** [reply]
I just tested it now but i found it inconvenient and slowing my work processus
Ok i guess i have to do with separate dedicated Sprites arf it would have been too easy if i was able to do it all with tilemaps haha
Concerning how tile data is stored i assume there is no possibility to increase it to add more data such as scale and rotation without impacting performences of tilemaps ?

**Reply by KoBeWi** [reply]
Each tile uses 64 bits to be stored in the TileMapLayer. The bits correspond to source id, tile id, alternate tile etc. It's stored like this, because ints are easy to serialize and bitwise operations are cheap.
Storing more data means either using bigger integers (more memory per tile) or adding some sort of associated structs for tiles. The latter *might* actually happen at some point, because it would allow to store per-tile color modulation and other things that people request. But I wouldn't count on it being implemented, at least not soon.

**Reply by Xinart** [reply]
Hoooo that was a clear, understanble and detailled answer thank you i think i understand more how tiles and tilemaps behind the scenes
So if my problem solution is needing more data on tiles this will note be for soon then good to know

**Comment by Xinart** [comment]
As recomended by @AThousandShips , i created an issue from this discussion as it sems to be a better place for this : https://github.com/godotengine/godot-proposals/issues/10308

**Comment by AThousandShips** [comment]
Closing in favour of the proposal issue
